addon/components/base/bs-dropdown.js

```diff
@@ -272,61 +283,88 @@ export default class Dropdown extends Component {
   @action
   toggleDropdown() {
     if (this.get('isOpen')) {
-      this.send('closeDropdown');
+      this.closeDropdown();
     } else {
-      this.send('openDropdown');
+      this.openDropdown();
     }
   }
 
   @action
   openDropdown() {
     this.set('isOpen', true);
-    this.addClickListener();
     this.get('onShow')();
   }
 
   @action
   closeDropdown() {
     this.set('isOpen', false);
-    this.removeClickListener();
     this.get('onHide')();
   }
 
-  addClickListener() {
-    if (!this.clickListener) {
-      this.clickListener = bind(this, this.closeOnClickHandler);
-      document.addEventListener('click', this.clickListener, true);
-    }
-  }
-
-  removeClickListener() {
-    if (this.clickListener)  {
-      document.removeEventListener('click', this.clickListener, true);
-      this.clickListener = null;
-    }
-  }
-
-  willDestroyElement() {
-    super.willDestroyElement(...arguments);
-    this.removeClickListener();
-  }
-
   /**
    * Handler for click events to close the dropdown
    *
    * @method closeOnClickHandler
    * @param e
    * @protected
    */
-  closeOnClickHandler(e) {
+  @action
+  closeHandler(e) {
     let { target } = e;
     let { toggleElement, menuElement } = this.getProperties('toggleElement', 'menuElement');
 
     if (!this.get('isDestroyed')
-      && (toggleElement && !toggleElement.contains(target))
-      && ((menuElement && !menuElement.contains(target)) || this.get('closeOnMenuClick'))) {
-      this.send('closeDropdown');
+      && (
+        (e.type === 'keyup' && e.which === TAB_KEYCODE && menuElement && !menuElement.contains(target))
+        || (e.type === 'click' && toggleElement && !toggleElement.contains(target) && ((menuElement && !menuElement.contains(target)) || this.get('closeOnMenuClick')))
+      )) {
+      this.closeDropdown();
+    }
+  }
+
+  @action
+  handleKeyEvent(event) {
+    if (/input|textarea/i.test(event.target.tagName)
+      ? (
+        event.which === SPACE_KEYCODE
+        || event.which !== ESCAPE_KEYCODE && (event.which !== ARROW_DOWN_KEYCODE && event.which !== ARROW_UP_KEYCODE || this.get('menuElement').contains(event.target)))
+      : !SUPPORTED_KEYCODES.includes(event.which)) {
```

Comment:
That's a very complex condition. An inline comment explaining it, would be great.

Reply:
Yeah, some copy pasting basically. And got me confused as well, *multiple times* . during working on it 😂. Will do